Jingle/SCJingleConverter/composition.cpp

```diff
@@ -849,37 +859,137 @@ uint32_t Composition::getMemUsage() {
+    optimizedNotesL.clear();
+    optimizedNotesR.clear();
 
-    Voice* voice = nullptr;
-    switch (chan) {
-    case RIGHT:
-        if (voices.count(voiceStrR)) {
-            voice = &voices[voiceStrR];
+    uint32_t removed_notes = 0;
+
+    // Optimize Notes in Left Channel
+    if (voices.count(voiceStrL)) {
+        // Collect notes into single vector
+        for (uint32_t meas_idx = meas_start_idx; meas_idx <= meas_end_idx; meas_idx++) {
+            Measure& meas = voices[voiceStrL].measures[meas_idx];
+            for (uint32_t notes_idx = 0; notes_idx < meas.notes.size(); notes_idx++) {
+                optimizedNotesL.push_back(meas.notes[notes_idx]);
+            }
         }
-        break;
 
-    case LEFT:
-        if (voices.count(voiceStrL)) {
-            voice = &voices[voiceStrL];
+        // Combine ties
+        for (uint32_t notes_idx = optimizedNotesL.size()-1; notes_idx-- > 0; ) {
+            Note& note = optimizedNotesL[notes_idx];
+
+            if (note.tied && notes_idx < optimizedNotesL.size()-1) {
+                Note& tie = optimizedNotesL[notes_idx+1];
+
+                float note_frequency = note.frequencies.back();
+                float tie_frequency = tie.frequencies.back();
+                if (chordIdxL < note.frequencies.size()) {
+                    note_frequency = note.frequencies[chordIdxL];
+                }
+                if (chordIdxL < tie.frequencies.size()) {
+                    tie_frequency = tie.frequencies[chordIdxL];
+                }
+
+                if (note_frequency == tie_frequency) {
+                    note.length += tie.length;
+                    optimizedNotesL.erase(optimizedNotesL.begin()+notes_idx+1);
+                    removed_notes++;
+                }
+            }
         }
-        break;
-    }
 
-    if (voice) {
-        if (meas_start_idx >= voice->measures.size() || meas_end_idx >= voice->measures.size()) {
-            qDebug() << "Start or End Measure Index out of bounds for Channel in getNumNotes()";
-            return 0;
+        // Remove trailing the trailing rest (should only be one max)
+        if (optimizedNotesL[optimizedNotesL.size()-1].frequencies[chordIdxL] == 0.f) {
+            optimizedNotesL.erase(optimizedNotesL.begin()+optimizedNotesL.size()-1);
         }
+    }
 
+    // Optimize Notes in Right Channel
+    if (voices.count(voiceStrR)) {
+        // Collect notes into single vector
         for (uint32_t meas_idx = meas_start_idx; meas_idx <= meas_end_idx; meas_idx++) {
-            Measure& meas = voice->measures[meas_idx];
-            num_notes += meas.notes.size();
+            Measure& meas = voices[voiceStrR].measures[meas_idx];
+            for (uint32_t notes_idx = 0; notes_idx < meas.notes.size(); notes_idx++) {
+                optimizedNotesR.push_back(meas.notes[notes_idx]);
+            }
+        }
+
+        // Combine ties
+        for (uint32_t notes_idx = optimizedNotesR.size()-1; notes_idx-- > 0; ) {
```

Comment:
notes_idx should not be decremented in comparison part of for loop

Reply:
Which would fit better?

```cpp
for (uint32_t notes_idx = optimizedNotes.size()-1; notes_idx < optimizedNotes.size(); notes_idx--) {
    ...
```
Still kinda ugly, now it'll wrap around, but it still works. 

```cp
for (uint32_t notes_idx = optimizedNotes.size()-1; notes_idx > 0; ) {
    notes_idx--;
    ....
```
Or this *only* pulls the notes_idx-- down out of the comparison.

Currently leaning towards the second.

Reply:
First, the logic in your first code block doesn't match the others. Was it a mistake that the condition check part of the for loop is `notes_idx < optimizedNotes.size()` and not `notes_idx > 0` or am I missing something?

Second, I'm less worried with text wrap around (assuming I understand your "ugly" comment) and more about code readability in terms functionality. The ++/-- features in C++ is great, but when it gets put into parts of a statement where people might not usually look for it, my opinion is that that code is hard to read and error prone if anyone were to change it. So I would think something like the following would be acceptable:

`for (uint32_t notes_idx = optimizedNotes.size()-1; notes_idx > 0; notes_idx--) {`

Saying all that though, I also have an addendum.  Is it possible we can get to this code and optimizedNotesR.size() will be zero? In that case, since `notes_idx` is unsigned, we will get wrap around and out of bounds issues. Initial thought is to change the for loop to count up and use that count value to compute the access index. Thoughts?

Reply:
> First, the logic in your first code block doesn't match the others. Was it a mistake that the condition check part of the for loop is `notes_idx < optimizedNotes.size()` and not `notes_idx > 0` or am I missing something?

Sorry, wrapping around is referring to integer underflow. If I use > 0, the first two notes will never be optimized since the loop never runs when `notes_idx == 0`. Then the compiler immediately yelled at me for trying `notes_idx <= 0` on an unsigned int, which led me to this. In this code, `notes_idx < optimizedNotes.size()` will evaluate false when `notes_idx` underflows and becomes greater than optimizedNotes.size(), which is kinda ugly.

Totally get what you're saying about putting things in unexpected places.

`for (uint32_t notes_idx = optimizedNotes.size()-1; notes_idx > 0; notes_idx--) {` 

Has the problem I touched on earlier, the loop will never run when `notes_idx == 0` so the first note (first two, sorta) get ignored and never optimized. Moving the `notes_idx--` to be the first statement in the loop is functionally identical to the weird `notes_idx-- > 0` without being as bad for readability. It allows the rest of the loop to run while `notes_idx == 0`, without having to deal with that in the conditional, either by decrementing it in an unnatural place or unintuitively dealing with underflow.

> Saying all that though, I also have an addendum. Is it possible we can get to this code and optimizedNotesR.size() will be zero? In that case, since notes_idx is unsigned, we will get wrap around and out of bounds issues.

Can't quite tell. It checks to see if the current active voice exists, then collects all notes from all measures in that voice into `optimizedNotes`. Are voices guaranteed to have at least some notes in them, even rests? As long as it has something, even if it's 100% rests we're good to go.

> Initial thought is to change the for loop to count up and use that count value to compute the access index. Thoughts?

You say using the count value to *compute* the access index... are you saying the loop would count up while the actual index accessed would still be counting backwards or? I'm having a hard time wrapping my head around that lol.

Reply:
I was thinking something like the following maybe:

```cpp
for (uint32_t notes_cnt = 1; notes_cnt < optimizedNotesR.size(); notes_cnt++) {
    uint32_t notes_idx = optimizedNotesR.size() - notes_cnt;
    ...
```

In this way we don't have to worry about whether optimizedNotesR.size() can ever be zero, because if it is we will never enter the loop to begin with. 

Thoughts?

Reply:
I haven't tested it yet but my initial thought is that something like that would be immediately thrown off once we start pulling notes off the end as they're optimized out, unless we add something to adjust the count every time that happens. If the goal's just to be sure that it's not empty, seems like making a separate, explicit check for that would be clearer.

Reply:
If the goal is just to make it run at == 0, then use a signed integer as the index var. As long as the index isn't larger than INT_MAX, there won't be any weird wrapping, and you can add asserts to that effect if you are worried about it. Also since you intentionally don't want to process the very last node first (because you are comparing with the next node), you can start at `- 2` and remove the ```< size - 1``` check below.

```cpp
for (int notes_idx = optimizedNotes.size() - 2; notes_idx >= 0; notes_idx--) {
```